### PR TITLE
Implement generate_title method for Conversation class

### LINKED ISSUE #253: We need a way to generate title per conversation
We should migrate over the logic from https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/utils/conversation_summary.py#L79 to `Conversation` class.

Basically we should implement `.generate_title` method of the conversation. It should check if there's at least ONE user message in the `state.events`, if not it should throw errors.

If they do, `.generate_title` will use the initial user message to generate a new title. `generate_title` should accept an optional `LLM` as input arg. If it exists, it will use it to generate title. If NOT, it will use `self.agent.llm` to do it

### PR DESCRIPTION
**Pull request by xingyaoww**
## Summary

This PR implements the `generate_title` method for the `Conversation` class as requested in issue #253. The method generates conversation titles based on the first user message using LLM completion with fallback to simple truncation.

### Implementation
- **Added `generate_title` method** to `Conversation` class in `openhands/sdk/conversation/conversation.py`
- **LLM Integration**: Uses LLM completion to generate concise, meaningful titles from the first user message
- **Flexible LLM Support**: Accepts optional `llm` parameter, falls back to `self.agent.llm` if not provided
- **Input Validation**: Validates that at least one user message exists in conversation events
- **Error Handling**: Graceful fallback to message truncation when LLM fails or returns empty response
- **Configurable Length**: Supports `max_length` parameter with automatic truncation and ellipsis

### Testing
- **Comprehensive Test Suite**: Added `tests/sdk/conversation/test_generate_title.py` with 6 test cases
- **Coverage**: Tests basic functionality, error handling, custom LLM usage, max length constraints, and edge cases
- **Proper Mocking**: Uses LiteLLM types for accurate mocking of LLM responses

## Method Signature

```python
def generate_title(self, llm: Optional[LLM] = None, max_length: int = 50) -> str:
    """Generate a title for the conversation based on the first user message.
    
    Args:
        llm: Optional LLM to use for title generation. If not provided, uses self.agent.llm.
        max_length: Maximum length of the generated title.
        
    Returns:
        A generated title for the conversation.
        
    Raises:
        ValueError: If no user messages are found in the conversation events.
    """
```

## Usage Examples

```python
# Basic usage
title = conversation.generate_title()

# With custom LLM
custom_llm = LLM(model="gpt-3.5-turbo", api_key=SecretStr("key"))
title = conversation.generate_title(llm=custom_llm)

# With custom max length
title = conversation.generate_title(max_length=30)
```

## Testing

All tests pass:
- ✅ Basic title generation functionality
- ✅ Error handling when no user messages exist
- ✅ Fallback behavior when LLM fails
- ✅ Max length truncation
- ✅ Custom LLM parameter usage
- ✅ Empty LLM response handling

## Migration from OpenHands

This implementation migrates the core logic from [OpenHands conversation_summary.py](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/utils/conversation_summary.py#L79) as requested, adapting it to work with the agent-sdk's `Conversation` class structure.

Fixes #253

@xingyaoww can click here to [continue refining the PR](https://app.all-hands.dev/conversations/4b54418563c749c6b33b1f5df28be31c)

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/All-Hands-AI/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Base Image | Docs / Tags |
|---|---|---|
| golang | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |
| java | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |

**Pull (multi-arch manifest)**
```bash
docker pull ghcr.io/all-hands-ai/agent-server:17238f7-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-17238f7-python \
  ghcr.io/all-hands-ai/agent-server:17238f7-python
```

**All tags pushed for this build**
```
ghcr.io/all-hands-ai/agent-server:17238f7-golang
ghcr.io/all-hands-ai/agent-server:v1.0.0_golang_tag_1.21-bookworm_golang
ghcr.io/all-hands-ai/agent-server:17238f7-java
ghcr.io/all-hands-ai/agent-server:v1.0.0_eclipse-temurin_tag_17-jdk_java
ghcr.io/all-hands-ai/agent-server:17238f7-python
ghcr.io/all-hands-ai/agent-server:v1.0.0_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_python
```

_The `17238f7` tag is a multi-arch manifest (amd64/arm64); your client pulls the right arch automatically._

### CODE REVIEW

File: openhands/sdk/conversation/conversation.py [hunk 1]
```diff
@@ -262,6 +264,126 @@ def close(self) -> None:
...
+                return title
+            else:
+                logger.warning("LLM returned empty response for title generation")
+                # Fall back to simple truncation
+                return self._generate_fallback_title(first_user_message, max_length)
+
+        except Exception as e:
+            logger.warning(f"Error generating conversation title with LLM: {e}")
+            # Fall back to simple truncation
+            return self._generate_fallback_title(first_user_message, max_length)
```
Comment (xingyaoww) [inline openhands/sdk/conversation/conversation.py]: @openhands Can we move most of the logic here to a util function to simplify the logic in conversation.py?

File: openhands/sdk/conversation/title_utils.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                        text=(
+                            "You are a helpful assistant that generates concise, "
+                            "descriptive titles for conversations with OpenHands. "
+                            "OpenHands is a helpful AI agent that can interact "
+                            "with a computer to solve tasks using bash terminal, "
+                            "file editor, and browser. Given a user message "
+                            "(which may be truncated), generate a concise, "
+                            "descriptive title for the conversation. Return only "
+                            "the title, with no additional text, quotes, or "
+                            "explanations."
```
Comment (xingyaoww) [inline openhands/sdk/conversation/title_utils.py:96]: Read https://gitmoji.dev/ and come up with a list of rules that conversation title should starts with emoji.

For example, We could define a vocabulary like:
:art: Frontend
:gear: Backend
:test_tube: Tests
:zap: DevOps
:rocket: Deployment
:package: Dependencies
:cd: Database
:broom: Chores


But please refer to https://gitmoji.dev/ for a list of standardized emojis

File: openhands/sdk/conversation/conversation.py [hunk 3]
```diff
@@ -262,6 +264,27 @@ def close(self) -> None:
...
+            max_length: Maximum length of the generated title.
+
+        Returns:
+            A generated title for the conversation.
+
+        Raises:
+            ValueError: If no user messages are found in the conversation.
+        """
+        # Use provided LLM or fall back to agent's LLM
+        llm_to_use = llm or self.agent.llm
```
Comment (malhotra5) [inline openhands/sdk/conversation/conversation.py]: Could this cause a race condition if we run two separate completions with the same LLM? Both completion events will attempt to modify the LLMs metrics object?

Comment (malhotra5) [reply openhands/sdk/conversation/conversation.py]: Ideally we use the registry from #288 and assign a unique LLM + service_id

Still working on simplifying the PR though, its a bit of overhead to enforce that pattern

Comment (xingyaoww) [reply openhands/sdk/conversation/conversation.py]: yeah i think i'll revisit this once you got that in

**Comment by xingyaoww** [comment]
@OpenHands merge from main and resolve the conflicts

File: openhands/sdk/conversation/title_utils.py (hunk 2 above)
Comment (ryanhoangt) [reply openhands/sdk/conversation/title_utils.py:96]: I implemented this in the latest commit.

File: openhands/sdk/conversation/conversation.py (hunk 3 above)
Comment (ryanhoangt) [reply openhands/sdk/conversation/conversation.py]: Have we finalized on this? I think the idea is to create a new LLM instance and request completion from the LLM registry rather than using the agent's LLM right?

Comment (xingyaoww) [reply openhands/sdk/conversation/conversation.py]: yep! that's right

Comment (xingyaoww) [reply openhands/sdk/conversation/conversation.py]: but i think it is probably fine to keep this as is and fallback to agent llm when we don't have this extra instance. We can wire this extra instance in when we shipped the rpod

File: openhands/sdk/conversation/impl/remote_conversation.py [hunk 4]
```diff
@@ -523,6 +524,27 @@ def update_secrets(self, secrets: Mapping[str, SecretValue]) -> None:
...
+        Raises:
+            ValueError: If no user messages are found in the conversation.
+        """
+        # Use provided LLM or fall back to agent's LLM
+        llm_to_use = llm or self.agent.llm
+
+        return generate_conversation_title(
+            events=self._state.events, llm=llm_to_use, max_length=max_length
+        )
+
```
Comment (xingyaoww) [inline openhands/sdk/conversation/impl/remote_conversation.py:551]: Can we add a related route in agent_server and remote conversation will make request to that endpoint?

File: openhands/sdk/conversation/impl/remote_conversation.py [hunk 5]
```diff
@@ -523,6 +523,26 @@ def update_secrets(self, secrets: Mapping[str, SecretValue]) -> None:
...
+        Args:
+            llm: Optional LLM to use for title generation. If not provided,
+                 uses the agent's LLM.
+            max_length: Maximum length of the generated title.
+
+        Returns:
+            A generated title for the conversation.
+        """
+        # For remote conversations, delegate to the server endpoint
+        payload = {"max_length": max_length}
```
Comment (xingyaoww) [inline openhands/sdk/conversation/impl/remote_conversation.py]: Why are we not supporting passing the LLM here?

Comment (ryanhoangt) [reply openhands/sdk/conversation/impl/remote_conversation.py]: Ah my bad, I updated it in the latest commit.

File: openhands/agent_server/conversation_service.py [hunk 6]
```diff
@@ -255,6 +255,31 @@ async def get_event_service(self, conversation_id: UUID) -> EventService | None:
             raise ValueError("inactive_service")
         return self._event_services.get(conversation_id)
 
+    async def generate_conversation_title(
+        self, conversation_id: UUID, max_length: int = 50, llm: LLM | None = None
+    ) -> str | None:
```
Comment (enyst) [inline openhands/agent_server/conversation_service.py:260]: On a side note, I'm always a bit confused by `conversation service` vs `event service`. I understand the conversation service is for handling conversations, while event service is for handling a single conversation, that is, the events in a single conversation?

If so, ... we know the conversation ID, here, the auto-title is a request to be executed for a single conversation, ... in the code we get the `event service` for this conversation, and run on it... I guess my question is, is there a reason why we go through the service intended for multiple conversations, when we know already we need only the service for a single conversation?

Sorry if I'm missing something obvious, maybe we always do this indirection?

File: openhands/agent_server/conversation_service.py [hunk 7]
```diff
@@ -255,6 +255,31 @@ async def get_event_service(self, conversation_id: UUID) -> EventService | None:
...
+        if llm and llm_service_id:
+            try:
+                llm = event_service._conversation.llm_registry.get(llm_service_id)
+            except KeyError:
+                # If service_id not found, register it as a new LLM
+                event_service._conversation.llm_registry.add(llm)
+
+        loop = asyncio.get_event_loop()
+        title = await loop.run_in_executor(
+            None, event_service._conversation.generate_title, llm, max_length
```
Comment (enyst) [inline openhands/agent_server/conversation_service.py:279]: This looks a bit wonky, we use a private `_conversation` variable in the event service

@openhands Make the `_conversation` var public, we do seem to need to access it outside the event service.

Note that this PR has been merged. Make a new branch from main, and a new PR with the fix.

File: openhands/agent_server/conversation_service.py (hunk 6 above)
Comment (xingyaoww) [reply openhands/agent_server/conversation_service.py:260]: I think this is probably a @tofarr question here in the early design of these two service
